maven_iuvs/echelle.py

```diff
@@ -2630,43 +2697,102 @@ def line_fit_initial_guess(wavelengths, spectrum, H_a=20, H_b=170, D_a=80, D_b=1
-    # because it will be calculated as a constant offset from the H central line, per advice from Mike Stevens.
-    lambda_H_lya_guess = 121.567
-    # Account for cases with huge wavelength shift
-    if np.abs(wavelengths[np.argmax(spectrum)]-121.567) > 0.05:
-        lambda_H_lya_guess = wavelengths[np.argmax(spectrum)]
-
-    # Background initial guess: assume a form y = mx + b. If m = 0, assume a constant offset.
-    bg_m_guess = 0
-    bg_b_guess = np.median(spectrum)
-
-    # Now do IPH.
-    DN_IPH_guess = DN_H_guess * 0.05 # wild guess
-    lambda_IPH_lya_guess = 121.57 # Arbitary
-    width_IPH_guess = 0.005 # Arbitrary. TODO: Can pass in after calculating based on sc ephemeris
-
-    return [DN_H_guess, DN_D_guess, DN_IPH_guess, 
-            lambda_H_lya_guess, lambda_IPH_lya_guess, 
-            width_IPH_guess, 
-            bg_m_guess, bg_b_guess]
+    for i in range(rows):
+        spectrum = spectra[i, :]
+        if (np.argmax(spectrum) < H_a) or (np.argmax(spectrum) > H_b):
+            H_a = max(0, np.argmax(spectrum) - 50) # Ensure we don't wrap into a negative index.
+            H_b = min(np.argmax(spectrum) + 50, len(spectrum))
+            D_a = max(0, np.argmax(spectrum) - 30)
+            D_b = min(np.argmax(spectrum) - 10, len(spectrum))
+            # Now control for either of these both somehow being accidentally set to 0
+            if D_a == D_b == 0:
+                D_b += 20
+    
+        # Now integrate to get an initial area guess
+        DN_H_guess = sp.integrate.simpson(spectrum[H_a:H_b])
+        DN_D_guess = sp.integrate.simpson(spectrum[D_a:D_b])
+        
+        # central wavelength initial guess - go with the canonical value. There is no need to return a guess for D
+        # because it will be calculated as a constant offset from the H central line, per advice from Mike Stevens.
+        lambda_H_lya_guess = 121.567
+        # Account for cases with huge wavelength shift
+        if np.abs(wavelengths[np.argmax(spectrum)]-121.567) > 0.05:
+            lambda_H_lya_guess = wavelengths[np.argmax(spectrum)]
+
+        # Now do IPH.
+        DN_IPH_guess = DN_H_guess * 0.05 # wild guess
+        width_IPH_guess = (IPH_maxw + IPH_minw) / 2
+
+        # Background initial guess: assume a form y = mx + b. If m = 0, assume a constant offset.
+        bg_m_guess = 0
+        bg_b_guess = np.median(spectrum)
+
+        guesses[i, :] = [DN_H_guess, DN_D_guess, DN_IPH_guess, 
+                         lambda_H_lya_guess, lambda_IPH_lya_guess[i], 
+                         width_IPH_guess, 
+                         bg_m_guess, bg_b_guess]
+
+    return guesses
+
+
+def predict_IPH_linecenter(light_fits):
+    """
+    Exact location of upstream direction not completely certain. 
+    Not accounted for: decelreation of IPH across solar system, assumed to be the same.
+    """
+    load_iuvs_spice()
```

Comment:
I think this will clear and load the spice kernels every time this function is called... possibly greatly impacting performance. May be better to check whether the kernels are loaded and return an error if they are not, or put the contents of the function into a try/except block that asks the user to load_iuvs_spice() if a spice error is thrown.

Reply:
Could you possibly add a change for this? I'm still working on error control for problem files.

Reply:
We discussed in person and determined not to fix this now, but will keep an eye on it if it becomes a problem for us or future users.